File: src-tauri/src/pipeline/connection.rs
use crate::pipeline::packet::Packet;
use crate::pipeline::port::{InputPort, OutputPort, OutputPorts};
use miette::Result;

use super::Component;

#[derive(Debug)]
pub struct Connection {
    pub input: InputPort,
    pub outputs: OutputPorts,
}

impl Connection {
    pub fn new() -> Connection {
        Self {
            input: InputPort::new(),
            outputs: OutputPorts::default(),
        }
    }

    pub async fn receive(&mut self) -> Option<Packet> {
        self.input.receive().await
    }

    pub async fn send(&mut self, packet: Packet) -> Result<()> {
        self.outputs.send(packet).await
    }

    pub fn connect(
        src: &mut dyn Component,
        dst: &mut dyn Component,
        src_port: &str,
        dst_port: &str,
    ) {
        let src = src.connection();
        let dst = dst.connection();
        src.attach(src_port, dst.accquire(dst_port))
    }

    fn attach(&mut self, src: &str, port: OutputPort) {
        self.outputs.attach(src, port);
    }

    fn accquire(&mut self, port: &str) -> OutputPort {
        self.input.accquire(port)
    }
}

File: operation/pipeline/connection.go
package pipeline

type Connection struct {
	Input   InputPort
	Outputs OutputPorts
}

func NewConnection() *Connection {
	return &Connection{
		Input:   NewInputPort(),
		Outputs: NewOutputPorts(),
	}
}

func (c *Connection) Receive() (*Packet, error) {
	return c.Input.Receive()
}

func (c *Connection) Send(packet *Packet) error {
	return c.Outputs.Send(packet)
}

func (c *Connection) Connect(src, dst *Connection, srcPort, dstPort string) {
	src.Attach(srcPort, dst.Accquire(dstPort))
}

func (c *Connection) Attach(src string, port OutputPort) {
	c.Outputs.Attach(src, port)
}

func (c *Connection) Accquire(port string) OutputPort {
	return c.Input.Accquire(port)
}
